Shell/Shell.cpp

```diff
@@ -1359,6 +1360,69 @@ void Shell::custom_event(Core::CustomEvent& event)
             Core::EventLoop::current().post_event(*this, make<Core::CustomEvent>(ShellEventType::ReadLine));
         return;
     }
+
+    if (event.custom_type() == ChildEvent) {
+#ifdef SH_DEBUG
+        dbg() << "SIGCHLD!";
+#endif
+        auto& jobs = this->jobs;
+        Vector<u64> disowned_jobs;
+        // Workaround the fact that we can't receive *who* exactly changed state.
+        // The child might still be alive (and even running) when this signal is dispatched to us
+        // so just...repeat until we find a suitable child.
+        // This, of course, will mean that someone can send us a SIGCHILD and we'd be spinning here forever.
+        bool found_child = false;
+        for (auto& it : jobs) {
+            auto job_id = it.key;
+            auto& job = *it.value;
+            int wstatus = 0;
+#ifdef SH_DEBUG
+            dbgf("waitpid({}) = ...", job.pid());
+#endif
+            auto child_pid = waitpid(job.pid(), &wstatus, WNOHANG | WUNTRACED);
+#ifdef SH_DEBUG
+            dbgf("... = {} - {}", child_pid, wstatus);
+#endif
+
+            if (child_pid < 0) {
+                if (errno == ECHILD) {
+                    // The child process went away before we could process its death, just assume it exited all ok.
+                    // FIXME: This should never happen, the child should stay around until we do the waitpid above.
```

Comment:
This is true, this should never happen, but it is only true for the first call to `waitpid`. Once `waitpid` is used on a terminated process, it is removed from the process table and subsequent calls will return `ECHILD`

Reply:
The entry in the job table should be removed if `waitpid` succeeds once though, I don't think this path would be taken twice for any one job

Reply:
Ok, I guess I assumed it's somehow hitting this path. If it is then that would be a Kernel bug.

Reply:
I'll add an assertion for it on the shell side (once I resolve this failing test).

Reply:
We'll see if that assertion trips any time in the future :P